go/store/nbs/journal.go

```diff
@@ -171,8 +184,33 @@ func trueUpBackingManifest(ctx context.Context, root hash.Hash, backing *journal
 	return mc, nil
 }
 
+// IterateRoots iterates over the in-memory roots tracked by the ChunkJournal and passes the root and associated
+// timestamp to a callback function, |f|. If |f| returns an error, iteration is stopped and the error is returned.
+func (j *ChunkJournal) IterateRoots(f func(root string, timestamp time.Time) error) error {
+	j.mu.Lock()
+	roots := j.roots
+	rootTimestamps := j.rootTimestamps
+	length := len(roots)
+	j.mu.Unlock()
+
+	if len(roots) != len(rootTimestamps) {
```

Comment:
I think this needs to go inside the lock, it's a race otherwise. The locals vars are references to the slices, other threads could be appending them concurrently

Reply:
oh, good catch! yup, I'm trying to capture the slice references and the length in the mutex, so that length check has got to be protected as well. Also another good reason why some documentation explaining the intent of the locking would be helpful, like your comment above suggested. thank u!

Reply:
Ended up pulling out a helper method for this (mostly to ensure the mutex is always unlocked, through `defer`) in 5484c3f34aedcb117976e62614924c624af8cab3. That was also a good place to explain the intent of the locking.